### PR TITLE
C and C++ domains, change to new logging

### PR DESCRIPTION
**Pull request by jakobandersen**
Also add ``get_source_info()`` to ``SphinxDirective`` and ``SphinxRole``.

### Feature or Bugfix
- Refactoring

### CODE REVIEW

File: sphinx/util/docutils.py [hunk 1]
```diff
@@ -388,12 +392,13 @@ def config(self) -> "Config":
...
 
-    def set_source_info(self, node: Node, lineno: int = None) -> None:
+    def get_source_info(self, lineno: int = None) -> Tuple[str, int]:
         if lineno is None:
             lineno = self.lineno
+        return self.inliner.reporter.get_source_and_line(lineno)  # type: ignore
 
-        source_info = self.inliner.reporter.get_source_and_line(lineno)  # type: ignore
-        node.source, node.line = source_info
+    def set_source_info(self, node: Node, lineno: int = None) -> None:
```
Comment (tk0miya) [inline sphinx/util/docutils.py:400]: Looks good. Could you introduce this helper in CHANGES also?

**Review by tk0miya — APPROVED** [review]
LGTM!

File: sphinx/domains/cpp.py [hunk 2]
```diff
@@ -4409,8 +4408,12 @@ class DefinitionParser(BaseParser):
 
     _prefix_keys = ('class', 'struct', 'enum', 'union', 'typename')
 
-    def __init__(self, definition: Any, warnEnv: Any, config: "Config") -> None:
-        super().__init__(definition, warnEnv)
+    def __init__(self, definition: str, *,
+                 location: Union[nodes.Node, Tuple[str, int]],
+                 emitWarnings: bool,
```
Comment (tk0miya) [inline sphinx/domains/cpp.py]: I'll try to suppress `emitWarning` parameter as a refactoring.

Comment (tk0miya) [reply sphinx/domains/cpp.py]: I just made a simple helper: #7327

File: sphinx/util/docutils.py (hunk 1 above)
Comment (jakobandersen) [reply sphinx/util/docutils.py:400]: Sure, done.

File: sphinx/domains/cpp.py (hunk 2 above)
Comment (jakobandersen) [reply sphinx/domains/cpp.py]: Awesome! This simplifies things a lot. Thanks!

**Pull request merged**